### PR TITLE
RSPEC-2974 : Classes without "public" constructors should be "final"

### PR DESCRIPTION
**Pull request by axel3rd**
Hi, 

Proposition for [RSPEC-2974](http://jira.sonarsource.com/browse/RSPEC-2974).

Two remarks : 
1) If I don't know how to configure ``core.autocrlf=true`` in Egit, please answer "WTF, I asked you in previous PR" :).
2) I have implemented the ``protected`` constructor behavior like [checkstyle:FinalClass](http://checkstyle.sourceforge.net/config_design.html#FinalClass) (no issue if one protected constructor) ; I could update if wanted

Best regards

### CODE REVIEW

**Comment by benzonico** [comment]
Thanks @axel3rd for your contribution. This looks quite good, would you be able to squash all your commits into one. I'll try to merge this within this week.

**Comment by axel3rd** [comment]
Thanks for feedback.
Good catch for the ``break``, I delete it with my last commit (``protected`` constructor imply an inverse logic).
I update code now and will do the squash ASAP (I need some learn about that).

File: java-checks/src/test/files/checks/FinalClassCheck.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+  public static int magic(){
+    return 42;
+  }
+}
+
+
+public class PrivateConstructorNotFinalClass {  // Noncompliant {{Make this class "final" or add a public constructor.}}
+
+  private PrivateConstructorClass() {
```
Comment (benzonico) [inline java-checks/src/test/files/checks/FinalClassCheck.java]: We do have a nice parser: this is not a constructor, it should be renamed.

File: java-checks/src/test/files/checks/FinalClassCheck.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  }
+
+  public static int magic(){
+    return 42;
+  }
+}
+
+public class PrivateAndPublicConstructorClass { // Compliant, one public constructor
+
+  private PrivateConstructorClass() {
```
Comment (benzonico) [inline java-checks/src/test/files/checks/FinalClassCheck.java]: We do have a nice parser: this is not a constructor, it should be renamed.

File: java-checks/src/test/files/checks/FinalClassCheck.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  }
+
+  public static int magic(){
+    return 42;
+  }
+}
+
+public class ProtectedConstructorClass {  // Compliant, all constructors are not private
+
+  protected PrivateConstructorClass() {
```
Comment (benzonico) [inline java-checks/src/test/files/checks/FinalClassCheck.java]: We do have a nice parser: this is not a constructor, it should be renamed.

File: java-checks/src/main/java/org/sonar/java/checks/FinalClassCheck.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  @Override
+  public List<Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.CLASS);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    ClassTree classTree = (ClassTree) tree;
+
+    // Check if all constructors are private
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/FinalClassCheck.java]: I don't think this comment is needed.

**Comment by benzonico** [comment]
Please see my latest (minor) comments. Otherwise it looks good, tell me on this conversation when you have updated the PR and I will merge it.

**Comment by axel3rd** [comment]
PR updated. Sorry for my poor previous push.

**Comment by benzonico** [comment]
Merged in https://github.com/SonarSource/sonar-java/commit/471735f373513b5aa054d3dec6a46fd6d0e3c949

**Comment by benzonico** [comment]
FYI @axel3rd I added the testcase of default constructor.

**Comment by axel3rd** [comment]
Thanks @benzonico

**Pull request closed without merging**